Plugin: Remove replace_editor filter, extend core editor

aduth opened the pull request:
Related: #11015

This pull request seeks to eliminate Gutenberg's `replace_editor` feature, instead allowing the default core block editor to load, changing Gutenberg's behavior to instead extend the core block editor and replace the core script handles with the Gutenberg latest versions.

Tasks separated out:

- [x] Deprecate / remove `gutenberg_add_admin_body_class` (#13572)
- [x] Defer vendor scripts to core. (#13573)
- [x] Drop inline scripts in favor of override_script improvements. (#13581)
- [x] Load locale data. (#12559)
- [x] Remove Gutenberg's heartbeat proxying. (#13576)
- [x] Apply Gutenberg's built `editor-styles.css` to editor settings by filter. (#13625)
- [x] Extract wp-editor-font stylesheet override (#14176)
- [x] Remove deprecated `_wpLoadGutenbergEditor`, `gutenberg` theme supports (#14144)
- [x] Await E2E promise interactions (#14219)

Needs confirmation as core aligned:

- [x] ~In fixing #5667, Gutenberg filters `user_can_richedit` as always true in the context of the block editor. Was this implemented in core?~
   - It was not implemented, but as discussed at https://github.com/WordPress/gutenberg/pull/5670#issuecomment-459043437 it did not need to be. See #13608 for removal from Gutenberg.
- [x] ~Does Gutenberg still need to extend / replace the `styles` property from the block editor settings?~
   - It's still necessary, as this is a build distributable from Gutenberg, and may need to change in [source](https://github.com/WordPress/gutenberg/blob/master/packages/editor/src/editor-styles.scss) to override that provided by core.
- [x] ~Gutenberg's heartbeat-to-hooks proxying must be implemented in core~
   - This was [already implemented](https://core.trac.wordpress.org/ticket/46107). Gutenberg's handling is redundant, and proposed for removal in #13576.
- [x] ~Confirm whether core implemented the "Screen Options" hiding equivalent to Gutenberg's `screen_options_show_screen` filtering~
   - As best I can tell, it did not, and should. See: https://core.trac.wordpress.org/ticket/46157

mcsf commented:
> * [ ]  In fixing #5667, Gutenberg filters `user_can_richedit` as always true in the context of the block editor. Was this implemented in core?

Commented in https://github.com/WordPress/gutenberg/pull/5670#issuecomment-459043437

aduth committed:
Plugin: Remove replace_editor filter, extend core editor

aduth committed:
Plugin: Avoid dynamic dependencies for wp-block-library style

This is handled by core in common blocks style enqueues behavior https://github.com/WordPress/wordpress-develop/blob/e421f26/src/wp-includes/script-loader.php#L2626-L2630

aduth commented:
Removing the "In Progress" label. Aside from #14144, I'm considering this as ready for review. I had some trouble with the end-to-end tests, specifically around the animations disabling. I think 44351ed should make this more durable.

The review should essentially be verifying that anything we were previously doing is respected still. Most of the additional filtering was extracted to separate pull requests. Most of the removed code should now be represented otherwise as part of the default [`edit-form-blocks.php` file](https://github.com/WordPress/wordpress-develop/blob/master/src/wp-admin/edit-form-blocks.php) which we now allow to load.

aduth committed:
Testing: Leverage saveDraft for meta boxes save

aduth commented:
Build is green now 🎉 

For what it's worth, I was having some issues with intermittent failures for the new tags end-to-end tests introduced in #13129 (@jorgefilipecosta). It should have been improved with #14219, so I might think it could have just been some caching problem.

aduth commented:
@jorgefilipecosta Can you please review this in mind of rebased refactor of widgets interoperability, per changes of #13511.

jorgefilipecosta commented on lib/client-assets.php:
```diff
@@ -299,7 +299,7 @@ function gutenberg_register_scripts_and_styles() {
 	gutenberg_override_style(
 		'wp-block-library',
 		gutenberg_url( 'build/block-library/style.css' ),
-		current_theme_supports( 'wp-block-styles' ) ? array( 'wp-block-library-theme' ) : array(),
```
I'm not understanding the reason for this change. I think 'wp-block-styles' should still depend on 'wp-block-library-theme' if the theme supports them.
gutenberg_override_style calls wp_register_style with the new dependencies so even if this dependency was set in core as we are registering it again we still need to pass it.

youknowriad replied:
I have vague memories that it's done in a different way in Core and this might align both approaches. The thing is, it's not possible to call `current_theme_supports` in the script loader in Core.

aduth replied:
Yes, there's a bit more of an explanation in in the extended comment description for https://github.com/WordPress/gutenberg/pull/13569/commits/2282ee4143db1d2f6ee01327317813446a4b2a8e . It still exists, but it's manually enqueued in core's implementation of the block scripts and styles function:

https://github.com/WordPress/wordpress-develop/blob/e421f26/src/wp-includes/script-loader.php#L2626-L2630

You can confirm this adding the following snippet somewhere in a theme which adds support for `wp-block-styles` (e.g. Twentynineteen):

```php
add_action( 'enqueue_block_assets', function() {
	var_export( current_theme_supports( 'wp-block-styles' ) );
	var_export( wp_style_is( 'wp-block-library-theme', 'enqueued' ) );
	exit;
} );
```

Should display "truetrue" on the editor screen.

jorgefilipecosta replied:
Thank you for the clarification @youknowriad, and @aduth.

jorgefilipecosta reviewed (commented):
I did a set of tests with legacy widgets (including JavaScript ones), and I did not notice any difference of behavior in this version when compared with the previous one. Thank you for applying these changes that isolate the legacy widget mechanism 👍 

I also did some tests with metaboxes and everything worked as before.

I left a comment regarding a dependency removal that I did not understand but other than that the changes look good to me.

aduth commented:
I did another force-push since it seemed Travis became stuck again.

jorgefilipecosta reviewed (approved):
In think we should merge this soon so we have more time to test. In my tests I did not found any regression and the changes make sense to me (althougth I dont' have deep knowlodge in some of the areas).

Thank you for this changes and improvements 👍

aduth commented:
Thanks for the reviews @jorgefilipecosta 👍
